WT-3310 Add log config option for alter.

sueloverso opened the pull request:
Not yet ready for review.  Basic code and test working but want to add more to the test.

sueloverso commented:
@michaelcahill and @agorrod The changes for session->alter to alter the log setting are ready for review.  There are a couple things to note:
1.  The new test uncovered a bug in the existing alter code for LSM where, although the lsm tree and existing chunks get updated in the metadata, a newly created chunk would get the original setting.  2.  I loosened the constraints on a `log:` cursor.  It now allows a log cursor to be opened if logging is not currently configured but log files exist (similar to printlog, but printlog takes a different path).  Doing so meant I refactored some of `log_scan`.
3.  Currently the alter code returns an error if the user tries to change the logging setting and a dhandle is open to that URI.  We might make that more generic for any alter setting.

agorrod commented on src/schema/schema_alter.c:
```diff
@@ -29,6 +30,36 @@ __alter_file(WT_SESSION_IMPL *session, const char *uri, const char *newcfg[])
+		 * we expect a return of 0.
+		 */
+		if (ret != WT_NOTFOUND) {
+			if (ret == 0)
```
Could we try to close the handle if it's open? I'm a bit wary of just giving an error message here, since the user has no way to ensure that the handle is closed.

sueloverso replied:
Recap of our discussion.  It may be not too difficult to force a close of a clean handle, perhaps some version of a forced sweep of a URI that solves this.  It may need refactoring of the sweep code or at least some investigation.  I created WT-3360 for this.

michaelcahill replied:
I guess I'm confused about why alter is different in this regard to, say, salvage.

Can't we request the handle with `WT_DHANDLE_EXCLUSIVE` (as we would for salvage), which will force it closed if it is open, or fail if a cursor is open?  Once alter releases the handle, any future access will reopen it and see the new settings.

sueloverso replied:
Ah!  Thanks.  I didn't realize that the exclusive flag already did all the things I wanted for this without complication.  I will close WT-3360 and do that work.

sueloverso commented:
@michaelcahill and @agorrod this PR is ready for review.  It now handles altering the log setting and will close handles.  There are a few things to note here:
1.  I added `WT_BTREE_ALTER` even though the operation is not specifically doing something to the btree structure.  There are two reasons I added it there instead of as `WT_DHANDLE_ALTER`.
   A.  It needs the infrastructure and handling of being one of the `BTREE_SPECIAL_FLAGS`.
   B.  There is only 1 bit left for a new DHANDLE flags and this would have used up the bitspace there.
2.  However, in doing so, `wt_lsm_tree_worker` needs to look at that flag and fix the metadata for the LSM tree itself and reread that.  So there is a layering violation there.
3.  Now that the alter code uses `wt_schema_worker` most of the code in `schema_alter.c` has been removed.  There's only one function in there.  But it didn't seem to fit in the btree code so if you have suggestions about where it would better live then we could get rid of that file.
4.  Since `alter` now always closes handles, the code in `test/format` needs to retry cursor opens.

agorrod commented on src/lsm/lsm_tree.c:
```diff
@@ -851,45 +851,6 @@ __wt_lsm_tree_retire_chunks(WT_SESSION_IMPL *session,
-	locked = true;
-
-	/* Alter the chunks. */
-	for (i = 0; i < lsm_tree->nchunks; i++) {
```
I'm surprised this step can be removed - can we talk this through?

sueloverso replied:
Yes. We'll talk.  But the short answer is that this is all now happens in `lsm_tree_worker` as it goes through all the chunks.

agorrod commented on src/schema/schema_alter.c:
```diff
@@ -48,122 +56,6 @@ __alter_file(WT_SESSION_IMPL *session, const char *uri, const char *newcfg[])
-	 * index or column group explicitly.
-	 */
-	if (table->ncolgroups == 0)
-		for (i = 0; i < WT_COLGROUPS(table); i++) {
```
Similarly to the LSM case - I'd have thought we would still need to apply the alter to each of the constituents of a table that has non-trivial schema.

sueloverso replied:
Similarly, `wt_schema_worker` does all this same work for us and gets us through each of the constituents of the table.

sueloverso commented:
autobuild: retest this please

sueloverso commented:
@agorrod This is ready for another review and merge.  I refactored some LSM code to free memory when needing to reread the metadata.  Please give that a look.  The Jenkins test-format-stress-sanitizer job has been running this branch all day today.

agorrod commented on src/lsm/lsm_tree.c:
```diff
@@ -15,15 +15,49 @@ static int __lsm_tree_open(
 
 /*
+ * __lsm_tree_meta_discard --
+ *	Free the metadata related LSM tree pointers.
```
I'm too familiar with thinking about metadata as the content of the metadata table - could you call this function something else? Maybe `__lsm_tree_discard_state`?

sueloverso replied:
Fixed.

agorrod commented on src/lsm/lsm_tree.c:
```diff
@@ -1385,6 +1358,22 @@ __wt_lsm_tree_worker(WT_SESSION_IMPL *session,
+		 * we just wrote.  Free the old ones.
+		 */
+		__lsm_tree_meta_discard(session, lsm_tree);
+		WT_ERR(__wt_lsm_meta_read(session, lsm_tree));
```
Any error from here must need to be fatal - right? Otherwise the in-memory tree structure will be invalid.

sueloverso replied:
Fixed.

agorrod commented on src/session/session_api.c:
```diff
@@ -149,6 +149,10 @@ __session_alter(WT_SESSION *wt_session, const char *uri, const char *config)
 	SESSION_API_CALL(session, alter, config, cfg);
 
+	/* In-memory ignores alter operations. */
+	if (F_ISSET(S2C(session), WT_CONN_IN_MEMORY))
```
I'm OK with this, since it applies to all the configuration options currently exposed via alter - I expect it to change in the future when we expose an alter option that makes sense for in-memory configuration. We can resolve that when it arises.

sueloverso commented on src/include/wiredtiger.in:
```diff
@@ -921,6 +921,12 @@ struct __wt_session {
 	 * cache.  Not compatible with LSM tables; see @ref
 	 * tuning_cache_resident for more information., a boolean flag; default
 	 * \c false.}
+	 * @config{log = (, the transaction log configuration for this object.
```
There is no change to any current behavior or default setting.  We don't document most API changes in there.  So my inclination is that it does not.  Let me know if you disagree.

agorrod commented:
Thanks @sueloverso
